Feature/127 add alert messages to dataset view

ekraffmiller opened the pull request:
## What this PR does / why we need it:
Adds Alert messages to Dataset View Page (for draft version, unpublished version, and version not found)

## Which issue(s) this PR closes:

- Closes #127 

## Special notes for your reviewer:

## Suggestions on how to test this:
View the alert messages in Storybook, in the DatasetAlerts section.  In the local dev environment, go to Dataverse and create the 3 scenarios (draft version, unpublished version, and version not found), then view the dataset page in the SPA.

## Does this PR introduce a user interface change? If mockups are available, please link/include them here:
no
## Is there a release notes update needed for this change?:
no
## Additional documentation:

MellyGray commented on public/locales/en/dataset.json:
```diff
@@ -22,8 +22,17 @@
+    },
+    "unpublishedDataset": {
+      "heading": "Unpublished Dataset Private URL",
+      "alertText": "Privately share this dataset before it is published: {{privateUrl}}"
```
This message "Privately share this dataset before it is published: {{privateUrl}}" is not being displayed in the SPA

JSF:
![image](https://github.com/IQSS/dataverse-frontend/assets/23359572/eddc79b5-8c74-40db-a1a7-0dc8f2ee897b)

SPA:
![image](https://github.com/IQSS/dataverse-frontend/assets/23359572/5afe642a-0b28-47ad-be17-f10308cbb32b)

ekraffmiller replied:
In the JSF, this message isn't displayed when the dataset is in Anonymized view, only when the dataset owner is in the normal view, after the privateUrlToken was created.  So I think we need to add a property to Dataset for the created token.  This would be a separate property from the privateUrlToken that is passed through searchParams.
I don't know if I'm explaining that clearly, so we can talk about it Monday. :)

MellyGray replied:
If I understood you well, we need js-dataverse to return a new Dataset property, the privateUrlToken, so it can be displayed in this alert from the normal view. Sounds good to me 👍

MellyGray reviewed (changes requested):
The component looks good but there are some behaviours that are different from JSF, take a look at my comments

ekraffmiller commented on public/locales/en/dataset.json:
```diff
@@ -22,8 +22,17 @@
+      "heading": "This draft version needs to be published",
+      "alertText": "When ready for sharing, please <b>publish</b> it so that others can see these changes"
+    },
+    "requestedVersionNotFound": {
```
This might be a case where we want to do something different from JSF - I think it makes sense to always show the alert if the version requested doesn't exist. I can bring it up with the team.

ekraffmiller replied:
We talked about it in the frontend meeting and decided to always show the alert, even if the only version is draft.  I added another translation for the slightly different text of this message.

ekraffmiller commented on public/locales/en/dataset.json:
```diff
@@ -22,8 +22,17 @@
-    "draftVersion": "<b>This draft version needs to be published</b>. When ready for sharing, please <b>publish</b> it so that others can see these changes",
-    "requestedVersionNotFound": "Info – Version {0} was not found. This is version {1}",
-    "unpublishedDataset": "Privately share this dataset before it is published: {0}"
+    "draftVersion": {
```
ok, I added a top margin the the Alerts container.

MellyGray replied:
the top margin looks good, but now the alert is not filling the whole space at the right 

![image](https://github.com/IQSS/dataverse-frontend/assets/23359572/5a939f29-9ff6-4b94-85f4-614ba85bd4d6)

ekraffmiller commented on src/stories/dataset/Dataset.stories.tsx:
```diff
@@ -13,6 +13,7 @@ import { WithSettings } from '../WithSettings'
 import { WithFilePermissionsGranted } from '../files/file-permission/WithFilePermissionsGranted'
+import { DatasetMockDraftVersionRepository } from './DatasetMockDraftVersionRepository'
 
 const meta: Meta<typeof Dataset> = {
```
I added a new method to DatasetMother to create a Dataset with a privateUrl.

ekraffmiller commented on public/locales/en/dataset.json:
```diff
@@ -22,8 +22,17 @@
-    "draftVersion": "<b>This draft version needs to be published</b>. When ready for sharing, please <b>publish</b> it so that others can see these changes",
-    "requestedVersionNotFound": "Info – Version {0} was not found. This is version {1}",
-    "unpublishedDataset": "Privately share this dataset before it is published: {0}"
+    "draftVersion": {
```
I fixed this, added logic that used the datasetPermissions in withAlerts().  In the current branch, all the dataset permissions are set to true (because they aren't returned from the API), so it still appears in the SPA, but it is not there in the Dataset Anonymized View Story.
<img width="1123" alt="Screenshot 2023-10-21 at 4 37 10 PM" src="https://github.com/IQSS/dataverse-frontend/assets/675224/dbc0dea4-6f1f-49bd-b5a1-5b447219dbf8">

ekraffmiller commented:
Hi @MellyGray , welcome back from vacation!
 I think I addressed all the issues, the only remaining one being needing to add privateUrlToken to the Dataset object, as described above.  If we agree that another property should be added, then I can update the code to use it.

Also, while testing on Demo dataverse, I noticed use case I hadn't seen before, 'Incomplete Metadata': 
<img width="1156" alt="Screenshot 2023-10-21 at 4 48 30 PM" src="https://github.com/IQSS/dataverse-frontend/assets/675224/c862d7dd-cc43-463b-95d6-1cd1203cbd82">

I'm not sure how the dataset got into that state - some required fields were missing. I think it may have gotten into a bad state when the Demo dataverse was updated - maybe a new metadata block was added?  But since the JSF has an alert for it, I think we need to add it (maybe in a subsequent ticket).  The style of it, with red text in a warning box, doesn't match our existing Alert component, so we should probably talk to the team about modifying it for the SPA.

MellyGray commented:
@ekraffmiller 

The applied changes look good! But now there is a space at the right which wasn't there before
![Captura de pantalla 2023-10-23 a las 13 02 42](https://github.com/IQSS/dataverse-frontend/assets/23359572/ddd5dbe1-4e68-47d2-8d19-6bd9ad9072f1)

Also, go ahead with adding privateUrlToken to the Dataset object

We can create a new issue for the Incomplete Metadata message and discuss the style with the team 👍

MellyGray reviewed (approved):
LGTM! Let's move this forward

kcondon commented:
Tested as recommended and I'm seeing the alerts in the running spa service. However, noticed a couple issues that are more about dataset state and navigation but related:
1. When the dataset is a draft and has never been published, ie no 1.0, the alert does say it is a draft and unpublished but does not show the unpublished tag under the title, only the draft tag. This is different from the current app
![Screen Shot 2023-10-26 at 4 50 50 PM](https://github.com/IQSS/dataverse-frontend/assets/4983816/14a52352-79ab-4711-8b39-cdcfc37d6dac)
![Screen Shot 2023-10-26 at 4 49 36 PM](https://github.com/IQSS/dataverse-frontend/assets/4983816/b471cb62-85c0-4193-976b-052f73385dac)

2. When referencing a non existent version, eg. 2.0 when there is only a 1.0 and a draft, in the old app is goes to 1.0 and provides the does not exist alert. In SPA, it goes to the latest version, ie. the draft, says 2.0 doesn't exist but this is the draft and needs to be published.
![Screen Shot 2023-10-26 at 4 58 02 PM](https://github.com/IQSS/dataverse-frontend/assets/4983816/c71a46fe-979a-446e-848c-6e22bd5c22b3)
![Screen Shot 2023-10-26 at 4 48 18 PM (1)](https://github.com/IQSS/dataverse-frontend/assets/4983816/d8477017-118f-45a9-8ae4-a9e47f3c9133)

ekraffmiller commented:
I fixed the logic for determining if the dataset has been published (this is temporary, and will be replaced with info from the server when integration is completed).  
For the requested version, if the version isn't found, it will only show the Draft if the user has permission to view the draft, otherwise it will show the latest published version.
